File: src/algo-old/3835.js
var countSubarrays = function (nums, k) {
  const minQueue = []
  const maxQueue = []
  let ans = 0
  let left = 0

  for (let right = 0; right < nums.length; right++) {
    const x = nums[right]
    while (minQueue.length > 0 && x <= nums[minQueue[minQueue.length - 1]]) {
      minQueue.pop()
    }
    minQueue.push(right)

    while (maxQueue.length > 0 && x >= nums[maxQueue[maxQueue.length - 1]]) {
      maxQueue.pop()
    }
    maxQueue.push(right)

    while ((nums[maxQueue[0]] - nums[minQueue[0]]) * (right - left + 1) > k) {
      left++
      if (minQueue[0] < left) {
        minQueue.shift()
      }
      if (maxQueue[0] < left) {
        maxQueue.shift()
      }
    }

    ans += right - left + 1
  }

  return ans
}

// console.log(countSubarrays([1, 3, 2], 4))
// console.log(countSubarrays([5, 5, 5, 5], 0))
// console.log(countSubarrays([1, 2, 3], 0))
console.log(countSubarrays([1000000000, 1, 1000000000, 1], 10000_0000_0000_000))

File: src/algo-1py/3835.py
class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        min_q = deque()
        max_q = deque()
        ans = left = 0
        for right, x in enumerate(nums):
            # 1. 入：元素进入窗口
            while min_q and x <= nums[min_q[-1]]:
                min_q.pop()
            min_q.append(right)

            while max_q and x >= nums[max_q[-1]]:
                max_q.pop()
            max_q.append(right)

            # 2. 出：如果窗口不满足要求，左端点离开窗口
            # 只需改下面这行代码，其余逻辑和 2762 题完全一致
            while (nums[max_q[0]] - nums[min_q[0]]) * (right - left + 1) > k:
                left += 1
                if min_q[0] < left:
                    min_q.popleft()
                if max_q[0] < left:
                    max_q.popleft()

            # 3. 更新答案
            ans += right - left + 1
        return ans
 